x-pack/plugins/apm/public/components/app/transaction_details/WaterfallWithSummmary/WaterfallContainer/index.tsx

```diff
@@ -29,9 +33,46 @@ export function WaterfallContainer({
     return null;
   }
 
+  const { legends, items } = waterfall;
+
+  // Service colors are needed to color the dot in the error popover
+  const serviceLegends = legends.filter(
+    ({ type }) => type === WaterfallLegendType.ServiceName
+  );
+  const serviceColors = serviceLegends.reduce((colorMap, legend) => {
+    return {
+      ...colorMap,
+      [legend.value!]: legend.color,
+    };
+  }, {} as Record<string, string>);
+
+  // only color by span type if there are only events for one service
+  const colorBy =
+    serviceLegends.length > 1
+      ? WaterfallLegendType.ServiceName
+      : WaterfallLegendType.SpanType;
+
+  const displayedLegends = legends.filter((legend) => legend.type === colorBy);
+
+  const legendsByValue = keyBy(displayedLegends, 'value');
+
+  items.forEach((item) => {
+    let color = '';
+    if ('legendValues' in item) {
+      color = legendsByValue[item.legendValues[colorBy]].color;
+    }
+
+    if (!color) {
+      // fall back to service color if there's no span.type, e.g. for transactions
+      color = serviceColors[item.doc.service.name];
+    }
+
+    item.color = color;
+  });
```

Comment:
nit: `map` > `forEach`

```suggestion
  const itemsWithColor = items.map((item) => {
    let color = '';
    if ('legendValues' in item) {
      color = legendsByValue[item.legendValues[colorBy]].color;
    }

    if (!color) {
      // fall back to service color if there's no span.type, e.g. for transactions
      color = serviceColors[item.doc.service.name];
    }

	return {...item, color };
  });
```

Reply:
I'm mutating objects because these items are referenced in other data structures as well (`childrenByParentId` and less relevant, `errorItems`), and I want to avoid rebuilding all of them. Not great, but trying to avoid bigger changes. Would a clarifying comment be good enough here?

Reply:
Ahh... okay that makes sense. Just leave it then. Hoping we can move to the shared component soon-ish.